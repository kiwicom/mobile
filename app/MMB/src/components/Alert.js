// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Text, StyleSheet, Icon } from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

type CommonProps = {|
  title: React.Element<typeof Translation>,
  children: React.Node,
|};

type Props = {|
  type: 'danger' | 'warning' | 'success',
  ...CommonProps,
|};

const AlertDanger = (props: CommonProps) => (
  <View style={[styles.container, styles.containerDanger]}>
    <View style={styles.iconContainer}>
      <Icon name="warning" size={18} color={RED} />
    </View>
    <View>
      <Text style={[styles.titleText, styles.titleDanger]}>{props.title}</Text>
      <View style={styles.body}>{props.children}</View>
    </View>
  </View>
);

export default function Alert(props: Props) {
  switch (props.type) {
    case 'danger':
      return <AlertDanger title={props.title}>{props.children}</AlertDanger>;
    case 'warning':
      // TODO
      return null;
    case 'success':
      // TODO
      return null;
    default:
      return null;
  }
}

const RED = '#d0021b';

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 15,
    paddingVertical: 11,
    flexDirection: 'row',
  },
  containerDanger: {
    backgroundColor: 'rgba(208, 2, 27, 0.15)',
  },
  iconContainer: {
    marginRight: 12,
  },
  titleText: {
    fontSize: 12,
  },
  titleDanger: {
    color: RED,
  },
  body: {
    marginTop: 5,
  },
});
