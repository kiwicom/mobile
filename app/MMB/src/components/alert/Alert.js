// @flow

import * as React from 'react';
import {
  StyleSheet,
  Icon,
  Color,
  type StylePropType,
} from '@kiwicom/mobile-shared';
import { Translation } from '@kiwicom/mobile-localization';

import RenderAlert from './RenderAlert';

type Props = {|
  +type: 'danger' | 'warning' | 'success',
  +title: React.Element<typeof Translation>,
  +children?: React.Node,
  +titleStyle?: StylePropType,
|};

const ICON_SIZE = 17;

export default function Alert(props: Props) {
  switch (props.type) {
    case 'danger':
      return (
        <RenderAlert
          icon={<Icon name="warning" size={ICON_SIZE} color={RED} />}
          title={props.title}
          titleStyle={[styles.titleDanger, props.titleStyle]}
          containerStyle={styles.containerDanger}
        >
          {props.children}
        </RenderAlert>
      );
    case 'warning':
      return (
        <RenderAlert
          icon={<Icon name="warning" size={ICON_SIZE} color="#eb9d08" />}
          title={props.title}
          titleStyle={[styles.titleWarning, props.titleStyle]}
          containerStyle={styles.containerWarning}
        >
          {props.children}
        </RenderAlert>
      );
    case 'success':
      return (
        <RenderAlert
          icon={
            <Icon
              size={ICON_SIZE}
              name="error-outline"
              color={Color.green.normal}
              style={styles.successIcon}
            />
          }
          title={props.title}
          titleStyle={[styles.titleSuccess, props.titleStyle]}
          containerStyle={styles.containerSuccess}
        >
          {props.children}
        </RenderAlert>
      );
    default:
      return null;
  }
}

const RED = '#d0021b';

const styles = StyleSheet.create({
  containerDanger: {
    backgroundColor: 'rgba(208, 2, 27, 0.15)',
  },
  successIcon: {
    transform: [{ rotate: '180deg' }],
  },
  containerSuccess: {
    backgroundColor: Color.green.light,
  },
  titleSuccess: {
    color: Color.green.dark,
  },
  containerWarning: {
    backgroundColor: 'rgba(235, 157, 8, 0.15)',
  },
  titleWarning: {
    color: '#eb9d08',
  },
  titleDanger: {
    color: RED,
  },
});
