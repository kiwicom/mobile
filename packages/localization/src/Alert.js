// @flow

import { Alert as AlertNative } from 'react-native';

import { type TranslationKeys } from './DefaultVocabulary';
import { type TranslationPromise } from './CancellableTranslation';
import { setTranslatedString, replaceValues } from './TranslationHelpers';

export type Translation =
  | {|
      +id: TranslationKeys,
      +values?: Object,
    |}
  | {|
      +passThrough: ?string | ?number,
    |};

type AlertButtons = Array<{
  text?: Translation,
  onPress?: ?() => void,
  style?: AlertButtonStyle,
}>;

type NativeAlertButtons = Array<{
  text?: string,
  onPress?: ?() => void,
  style?: AlertButtonStyle,
}>;

type AlertOptions = {
  cancelable?: ?boolean,
  onDismiss?: ?() => void,
};

type AlertButtonStyle = 'default' | 'cancel' | 'destructive';

type AlertType = 'default' | 'plain-text' | 'secure-text' | 'login-password';

const getTranslation = async (translation: ?Translation) => {
  let translatePromise: TranslationPromise | null = null;
  if (translation && translation.id) {
    const translatedTitle = await setTranslatedString(
      translation.id,
      translatePromise,
      translatedString => translatedString,
    );
    if (translatedTitle) {
      return replaceValues(translatedTitle, translation.values);
    }
  } else if (translation && translation.passThrough) {
    return String(translation.passThrough);
  }
  return undefined;
};

export default class Alert extends AlertNative {
  static translatedAlert = async (
    title: ?Translation,
    message?: Translation,
    buttons?: AlertButtons,
    options?: AlertOptions,
    type?: AlertType,
  ) => {
    const translatedTitleWithValues = await getTranslation(title);
    const translatedMessageWithValues = await getTranslation(message);

    let translatedButtons: NativeAlertButtons = [];
    if (buttons && buttons.length >= 1) {
      await Promise.all(
        buttons.map(async button => {
          if (button && button.text) {
            const translatedButtonText = await getTranslation(button.text);
            translatedButtons.push({
              onPress: button.onPress,
              style: button.style,
              text: translatedButtonText,
            });
          }
        }),
      );
    }

    AlertNative.alert(
      translatedTitleWithValues,
      translatedMessageWithValues,
      translatedButtons,
      options,
      type,
    );
  };
}
