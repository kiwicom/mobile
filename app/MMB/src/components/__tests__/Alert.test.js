// @flow

import * as React from 'react';
import { PlaygroundRenderer } from '@kiwicom/mobile-playground';
import { Translation } from '@kiwicom/mobile-localization';

import Alert from '../Alert';

describe('Alert', () => {
  it('renders danger type', () => {
    PlaygroundRenderer.render(
      <Alert type="danger" title={<Translation passThrough="Alert danger" />}>
        <Translation passThrough="Alert body" />
      </Alert>,
    );
  });
});
