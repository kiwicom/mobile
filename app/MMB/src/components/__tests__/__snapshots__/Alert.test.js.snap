// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders danger type 1`] = `
<AlertDanger
  title={
    <Translation
      passThrough="Alert danger"
    />
  }
>
  <Translation
    passThrough="Alert body"
  />
</AlertDanger>
`;
